src/main/java/kafka/vertx/demo/PeriodicProducer.java

```diff
@@ -29,15 +30,28 @@
 
   @Override
   public void start(Promise<Void> startPromise) {
-    ConfigRetriever.create(vertx,
+    loadKafkaConfig()
+      .onSuccess(config -> setup(config, startPromise))
```

Comment:
Use chaining here - it's not `setup`'s job to complete the promise.

```.map(config -> setup(config))
.onSuccess(startPromise::complete)
.onFailure(startPromise::fail);
```

If however you are on vert.x < 4 - I think you have to use onComplete the onFailure handler will override the onSuccess handler:

```.map(config -> setup(config))
.onComplete(ar -> ar.succeeded() ? startPromise.complete() : startPromise.fail(ar.cause()))
```